# 1 Replacing a recursion by a Dynamic Program
```

```



In [14]:
def recurse_original(n):
    print(n)
    if n > 1:
        recurse(n-1)
        recurse(n//2)
        
recurse_original(5)

5
4
3
2
1
1
1
2
1
1
2
1
1


* #### 1. Your good genie has provided you with the start of a Dynamic Program to replace the expensive function above. Just fill in the blanks below (or rather, dots)!

In [0]:
def recurse(n):
    print(n)
    if n > 1:
        recurse(n-1)
        recurse(n//2)

def dp(n):
    l = [None]*(n+1)
    l[0] = ""  #empty string
    for i in range(1, n+1):
        l[i] = str(i) + "\n" + l[i-1] + l[i//2]
    print(l[n])

In [6]:
recurse(5)

5
4
3
2
1
1
1
2
1
1
2
1
1


* #### 2. How would you change the function `dp` to count and output the number of lines that the function recurse outputs?

In [0]:
def count_recurse_lines(n):
    l = [None]*(n+1)
    l[0] = 0  #nothing to print out
    for i in range(1, n+1):
        l[i] = 1 + l[i-1] + l[i//2]
    print(l[n])
  

In [18]:
count_recurse_lines(5)

13


* #### 3. What is the time complexity of `dp` in $O()$?

$O(n)$

# 2 Recursion with memoisation to Dynamic Programming

The problem of counting the number of ways to write a positive integer n as a sum of positive numbers.

In [21]:
def ways(n, mem=None):
    calls = 1
    if mem is None:
        mem = [None] * (n+1)
    
    mem[n] = 1  # n = n
    for i in range(1, n):
        if mem[i] is None:
            (w, c) = ways(i, mem)
        else:
            w = mem[i]
            c = 0
        mem[n] += w
        calls += c
    return (mem[n], calls)

ways(4)

(8, 4)

* #### 1. We ask that you write a DP that constructs the list (mem) used for memoisation, but in an iterative manner.

In [0]:
def ways_dp(n, mem=None):
    if mem is None:
        mem = [None]*(n+1)
    mem[0] = 1
    
    for i in range(1, n+1):
        w = 1
        for j in range(1, i):
            w += mem[j]
        mem[i] = w
    return mem[n]

In [28]:
ways_dp(0)

1

* #### 2. What is the size of the state space of the DP you wrote? Is this surprising?

$O(n)$?

Each cell stores the number of ways representing a number by summing.

# 3 Optimal salesperson

Suppose that you are a door-to-door salesperson, selling the latest innovation in vacuum cleaners to less-than-enthusiastic customers. Today, you are planning on selling to home of the n houses along a particular street. You are a master salesperson, so for each house, you have already worked out the amount $c_i$ of profit that you will make from the person in house $i$ if you talk to them. Unfortunately, you cannot sell to every house, since if a person's neighbour sees you selling to them, they will hide and not answer the door for you. Therefore, you must select a subset of houses to sell to such that none of them are next to each other, and such that you make the maximum amount of money.

For example, if there are 10 houses and the profits that you can make from each of them are 50, 10, 12, 65, 40, 95, 100, 12, 20, 30, then it is optimal to sell to the houses 1, 4, 6, 8, 10 for a total profit of $252.

Obviously, the maximum selecting space is $\frac{N_{house}}{2}$.

Starting with a house, gain profit $p$, has 4 remaining selection.

* #### 1. Devise a dynamic programming to solve this problem.

In [0]:
def max_profit(profits):
    max_visit = len(profits)//2
    n = len(profits)
    f = [[0 for _ in range(max_visit+1)] for _ in range(n+1)]
        
    # No house is visited. Get 0 profit.
    for no_visit in range(max_visit+1):
        f[0][no_visit] = 0
    
    visited_flag = False
    for i in range(1, max_visit+1):
        for j in range(1, n+1):
            if visited_flag == True:
                f[j][i] = f[j-1][i]
                visited_flag = False
            else:
                new_profit1 = profits[j-1] + f[j-1][i]
                new_profit2 = profits[j] + f[j-1][i] if j <= n else 0
                f[j][i] = max(new_profit1, new_profit2)
                visited_flag = True
            
    return f[n][max_visit], f

In [93]:
max_p, f = max_profit([50,10,12,65,40,95,100,12,20,30])
print("Max profit: %d"%max_p)
for r in f:
    print(r)

Max profit: 340
[0, 0, 0, 0, 0, 0]
[0, 50, 50, 50, 50, 50]
[0, 50, 50, 50, 50, 50]
[0, 115, 115, 115, 115, 115]
[0, 115, 115, 115, 115, 115]
[0, 210, 210, 210, 210, 210]
[0, 210, 210, 210, 210, 210]
[0, 310, 310, 310, 310, 310]
[0, 310, 310, 310, 310, 310]
[0, 340, 340, 340, 340, 340]
[0, 340, 340, 340, 340, 340]


[50, 10, 12, 65, 40, 95, 100, 12, 20, 30]

[50, 50, 50, 115, 115, 210, 210, 222, 222, 252]

* #### 2. Extend your program so that it returns an optimal subset of houses to sell to, in addition to the maximum possible profit.

In [0]:
def visit_houses(profits):
    max_visit = len(profits)//2
    n = len(profits)
    f = [[0 for _ in range(max_visit+1)] for _ in range(n+1)]
        
    # No house is visited. Get 0 profit.
    for no_visit in range(max_visit+1):
        f[0][no_visit] = 0
    
    visited_flag = False
    for i in range(1, max_visit+1):
        for j in range(1, n+1):
            if visited_flag == True:
                f[j][i] = f[j-1][i]
                visited_flag = False
            else:
                new_profit1 = profits[j-1] + f[j-1][i]
                new_profit2 = profits[j] if j < n else 0
                f[j][i] = max(new_profit1, new_profit2)
                visited_flag = True
            
    # Build the visited record by difference
    visited = []
    for i in range(n+1):
        if f[i][-1] - f[i-1][-1] == 0:
            visited.append(i-1)
    return visited

In [89]:
visit_houses([50,10,12,65,40,95,100,12,20,30])

[2, 4, 6, 8, 10]

# 4. Distinct "number of ways" without repeat

We are now interested in the number of non-ordered ways to write a non-negative integer $n$ as a sum of distinct non-negative numbers, i.e. without repeats. For instance, there are only two ways to write the number 4: 1+3 and 4. Write a DP that has a state space of size $O(n^2)$.

In [0]:
def ways_nonrp_dp(n, mem=None):
    if n <= 2:
        return 1
    if mem is None:
        mem = [None]*(n+1)
        mem[0] = 0
        mem[1] = 1
        mem[2] = 1
    
    for i in range(3, n+1):
        j = (i-1)//2
        mem[i] = 1 + mem[j] + mem[j-1]
    return mem[n]

def foo(n):
    if n == 1:
        return 1
    elif n <= 0:
        return 0
    else:
        return 1 + foo((n-1)//2) + foo(((n-1)//2)-1)

In [14]:
for i in range(1, 25):
    #print(i, ' ', foo(i))
    print(i, ' ', ways_nonrp_dp(i))

1   1
2   1
3   2
4   2
5   3
6   3
7   4
8   4
9   5
10   5
11   6
12   6
13   7
14   7
15   8
16   8
17   9
18   9
19   10
20   10
21   11
22   11
23   12
24   12
